### PR TITLE
Support for Hamamatsu Streak tiff metadata

### PR DESCRIPTION
**Pull request by LMSC-NTappy**
This PR concerns users of streak camera for time-resolved luminescence. Thus, if either @jlaehne, @jordiferrero or @ericpre have time to take a look, it would be great.

A significant part of the changes is a refactoring of the scale-parsing functions for .tiff files obtained from different SEM,TEM and optical microscope manufacturers. Previously, all manufacturer-specific operations where handled in a single function. I have moved them to individual functions and kept the main `_parse_scale_unit` function as some sort of interface.

Tests pass on my side.

Cheers,

Nicolas

### Description of the change
A few sentences and/or a bulleted list to describe and motivate the change:
- Added axes and metadata parsing support for .tiff files generated by hamamatsu streak cameras
- Refactored the .tiff `_parse_scale_unit` function for clarity and ease of expanding support in the future
- Support initialization of non-uniform axes
- Issue warning in case of axis linearization error
- Support for a tiff-specific keyword argument in case of non-uniform axis parsing.

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x]  Updated Documentation (also for #2862, which was missing)
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import hyperspy.tests.io
import os

dirname = os.path.join(os.path.dirname(hyperspy.tests.io.__file__),"tiff_files")
file = 'test_hamamatsu_streak_SCAN.tif'
#Get the example streak map
fname = os.path.join(dirname, file)

s = hs.load(fname) #Load as a linear axes but warn about linearisation
#Load as DataAxis
s = hs.load(fname, hamamatsu_streak_axis_type= 'data')
#Load as FunctionalDataAxis (3rd degree polynomial)
s = hs.load(fname, hamamatsu_streak_axis_type= 'functional')
#Load as UniformDataAxis without warning
s = hs.load(fname, hamamatsu_streak_axis_type= 'uniform')  
```

### CODE REVIEW

**Comment by jlaehne** [comment]
A colleague has a matlab script to read the binary format for streak camera data (I think also Hamamatsu). Would be a good first IO-plugin for https://github.com/lumispy/lumispy . I would try to get a student to do the implementation, unless someone wants to give it a shot already.

File: hyperspy/io_plugins/tiff.py [hunk 1]
```diff
@@ -451,9 +438,188 @@ def wave_len(ht):
...
+    if not 'Software' in op:
+        is_hamatif = False
+        return is_hamatif
+    else:
+        software = op['Software']
+        if not software.startswith('HPD-TA'):
+            is_hamatif = False
+
+    return is_hamatif
+def _get_hamamatsu_streak_description(tiff, op):
```
Comment (jlaehne) [inline hyperspy/io_plugins/tiff.py:536]: blank line before new `def` (several occurences)

File: hyperspy/io_plugins/tiff.py [hunk 2]
```diff
@@ -451,9 +438,188 @@ def wave_len(ht):
...
+    for row in reader:
+        key = row[0].strip(" []")
+        key_dict = {}
+        for element in row[1:]:
+            spl = element.split('=')
+            if len(spl) == 2:
+                key_dict[spl[0]] = spl[1].strip('"')
+        dict_meta[key] = key_dict
+
+    import re
```
Comment (jlaehne) [inline hyperspy/io_plugins/tiff.py]: Imports in the middle of functions are quite unusual. Should at least be at the top of the method, but I think they are generally at the top of the files across HyperSpy,

**Comment by jlaehne** [comment]
Looks good on first sight. A bit hard to review due to the restructuring (but I also hope the tests do their jobs as far as they exist).

There are quite a number of codecov warnings (see 'Files changed' tab). Could you try to increase coverage further, even though most of them are for the code that you moved around?

File: hyperspy/io_plugins/tiff.py (hunk 2 above)
Comment (LMSC-NTappy) [reply hyperspy/io_plugins/tiff.py]: Ok, the idea was to run import only if the .tiff is recognized as hamamatsu but fair enough it doesn't matter much

File: hyperspy/io_plugins/tiff.py (hunk 1 above)
Comment (LMSC-NTappy) [reply hyperspy/io_plugins/tiff.py:536]: corrected (also other style improvements)

**Comment by LMSC-NTappy** [comment]
Thank you for reviewing, I have implemented your recommendations

> A colleague has a matlab script to read the binary format for streak camera data (I think also Hamamatsu). Would be a good first IO-plugin for https://github.com/lumispy/lumispy . I would try to get a student to do the implementation, unless someone wants to give it a shot already.

Always good to support more files I guess. On my side, working with .tiff files is enough so go ahead.

> Looks good on first sight. A bit hard to review due to the restructuring (but I also hope the tests do their jobs as far as they exist).


I understand, but in the long run I hope this helps further extending .tiff loading capabilities with a cleaner code.


> There are quite a number of codecov warnings (see 'Files changed' tab). Could you try to increase coverage further, even though most of them are for the code that you moved around?

The missing lines cover manufacturer-specific .tiff metadata (Zeiss, olympus etc), which I don't have access to. I could recreate it by altering another file with the tiff library I guess, but that would be for the sole purpose of making sure that the loader does not attempt to read metadata in very specific cases of invalid files (e.g, the manufacturer-expected field exists but it has not the correct value inside). I think we can trust the software to work as intended in those cases

**Comment by LMSC-NTappy** [comment]
I have one more question though. 

In the hamamatsu case I cover, there is a linearization of the time axis being performed. The non-linearized axes are put in the `original_metadata` field for the sake of preserving information. 

In that case, is the policy to raise a warning upon loading the file?

**Comment by jlaehne** [comment]
I could provide a Zeiss tiff file, if you would be willing to integrate that in the tests?

**Comment by LMSC-NTappy** [comment]
I don't think there are any uncovered lines regarding zeiss .tiffs in the codecov right now.

The unhandled cases are: 
- conversion of the tiff unit when it was saved as centimeters and loaded calling force_read_resolution : https://github.com/LMSC-NTappy/hyperspy/blob/18daa09f0c225c2085dfcf631f5acb250c48c9d6/hyperspy/io_plugins/tiff.py#L305-L308
- Missing `'FEI_HELIOS`' key in an FEI_SEM metadata file: https://github.com/LMSC-NTappy/hyperspy/blob/18daa09f0c225c2085dfcf631f5acb250c48c9d6/hyperspy/io_plugins/tiff.py#L322-L326
- Missing `'PixelSizeX'` key in a tvips file: https://github.com/LMSC-NTappy/hyperspy/blob/18daa09f0c225c2085dfcf631f5acb250c48c9d6/hyperspy/io_plugins/tiff.py#L373-L376
- Jeol_sightX file for unit set in cm and not inches: https://github.com/LMSC-NTappy/hyperspy/blob/18daa09f0c225c2085dfcf631f5acb250c48c9d6/hyperspy/io_plugins/tiff.py#L430-L434
- Jeol_sightX file and camera length set in cm and not mm: https://github.com/LMSC-NTappy/hyperspy/blob/18daa09f0c225c2085dfcf631f5acb250c48c9d6/hyperspy/io_plugins/tiff.py#L452-L453

For the unit conversion cases, I think we can trust that if it works correctly from inches/mm, it will also work for cm since it's just the factor that changes.

For my hamamatsu metadata, I don't test that the `_is_streak_hamamatsu` function correctly returns False if the `'Artist'` or `'Software'` keys of metadata do exist but contain values incompatible with a streak map. I can write test for it if you want but tbh I really don't think it's necessary

File: hyperspy/models/model1d.py [hunk 3]
```diff
@@ -436,7 +436,7 @@ def __call__(self, non_convolved=False, onlyactive=False,
             if self.signal.axes_manager[-1].is_uniform:
                 to_return *= self.signal.axes_manager[-1].scale
             else:
-                to_return *= np.gradient(self.signal.axes_manager[-1].axis)
+                to_return *= np.gradient(self.signal.axes_manager[-1].axis[self.channel_switches])
```
Comment (ericpre) [inline hyperspy/models/model1d.py]: This is unrelated to this PR, can you please do this change in a separate PR?

File: hyperspy/io_plugins/tiff.py [hunk 4]
```diff
@@ -254,6 +252,9 @@ def _read_serie(tiff, serie, filename, force_read_resolution=False,
     else:
         dc = _load_data(*data_args, memmap=memmap, **kwds)
 
+    if _is_streak_hamamatsu(op):
+        op.update({'ImageDescription_Parsed': _get_hamamatsu_streak_description(tiff, op)})
```
Comment (ericpre) [inline hyperspy/io_plugins/tiff.py]:
```suggestion
        op.update({'ImageDescriptionParsed': _get_hamamatsu_streak_description(tiff, op)})
```
Just to avoid mixing CamelCase and snake_case in the same name! ;)

**Review by ericpre — COMMENTED** [review]
> I have one more question though.
>
> In the hamamatsu case I cover, there is a linearization of the time axis being performed. The non-linearized axes are put in the original_metadata field for the sake of preserving information.
>
> In that case, is the policy to raise a warning upon loading the file?

This sounds like a very good opportunities to use a non-uniform axis?! :)

> For my hamamatsu metadata, I don't test that the _is_streak_hamamatsu function correctly returns False if the 'Artist' or 'Software' keys of metadata do exist but contain values incompatible with a streak map. I can write test for it if you want but tbh I really don't think it's necessary

If you think that it is not worth it adding a test, then please add `  # pragma: no cover`. In this case, this can be sorted by writing a simple test:  which could consist in opening a Hamamatsu file with tifffile, edit the tag, save the file and try to re-open it with hyperspy. Because it should be a simple test, I would be in favour of adding it.

Is it possible to have smaller test file? 600KB is not a big deal, but smaller would be better (10s of KB).

File: hyperspy/models/model1d.py (hunk 3 above)
Comment (LMSC-NTappy) [reply hyperspy/models/model1d.py]: Oups, sorry my bad

**Comment by LMSC-NTappy** [comment]
> This sounds like a very good opportunities to use a non-uniform axis?! :)

Indeed! I tend to forget that non-uniform axes will be released together with this patch :). I'll head this way but then I will need to pass **kwds to `_parse_scale_unit` so that the axis can be enforced linear. I am sure otherwise there will be some plot issues if one axis is linear and the other one is not.

> Is it possible to have smaller test file? 600KB is not a big deal, but smaller would be better (10s of KB).

I know, but I just cannot get anything smaller than that... I don't know how to crop/rebin/reduce their size without messing up the metadata. That's also why `_is_streak_hamamatsu` is so strict. If the file is altered everything seem to go south. You have a hint on how to do it maybe?

**Comment by LMSC-NTappy** [comment]
> If you think that it is not worth it adding a test, then please add # pragma: no cover

I just have added the tests in the end. Wasn't such a big deal. Let's see if coverage improves

**Comment by LMSC-NTappy** [comment]
Ok apparently the failing test are due to a deprecation in the tifffile library. 

I will fix that, but probably by removing the saving of .tiff altogether and manually passing the modified dictionaries to `_is_streak_hamamatsu`.  Otherwise it will get a bit messy

**Comment by ericpre** [comment]
> > This sounds like a very good opportunities to use a non-uniform axis?! :)
> 
> Indeed! I tend to forget that non-uniform axes will be released together with this patch :). I'll head this way but then I will need to pass **kwds to `_parse_scale_unit` so that the axis can be enforced linear. I am sure otherwise there will be some plot issues if one axis is linear and the other one is not.

Yes, since not all feature are supported with non-uniform axes, it is possible that plot doesn't work out of the box. An alternative is to convert manually after reading the data using `convert_to_uniform_axis`.

> > Is it possible to have smaller test file? 600KB is not a big deal, but smaller would be better (10s of KB).
> 
> I know, but I just cannot get anything smaller than that... I don't know how to crop/rebin/reduce their size without messing up the metadata. That's also why `_is_streak_hamamatsu` is so strict. If the file is altered everything seem to go south. You have a hint on how to do it maybe?

Ok, if this is not possible to acquire a smaller file, then this is not worth it.
If it uses compression, it may be worth trying acquiring an image with all values identical, as this is something that should compress well - obviously, this is assuming that the size of the data comes from the data itself and not the metadata!

**Comment by LMSC-NTappy** [comment]
> Yes, since not all feature are supported with non-uniform axes, it is possible that plot doesn't work out of the box. An alternative is to convert manually after reading the data using `convert_to_uniform_axis`.

I will give it a shot nevertheless, but there a bit more of refactoring involved so it will take a bit of time. 


> Ok,` if this is not possible to acquire a smaller file, then this is not worth it.
If it uses compression, it may be worth trying acquiring an image with all values identical, as this is something that should compress well - obviously, this is assuming that the size of the data comes from the data itself and not the metadata!

That's a very good point. Saving the file in compressed form reduces to 35 kB but so far I didn't have any luck in preserving the metadata. I will check next week with the tool administrator, who also has access to the software, if he can help me with this

Cheers

Nico

**Comment by LMSC-NTappy** [comment]
Ok, now implemented with support for reading non-uniform axes. Behaviour is as follow

```python
s = hs.load('example_hamamatsu.tiff') #Load as a linear axes but warn about linearisation
#Load as DataAxis
s = hs.load('example_hamamatsu.tiff',hamamatsu_streak_axis_type= 'data')
#Load as FunctionalDataAxis (3rd degree polynomial)
s = hs.load('example_hamamatsu.tiff',hamamatsu_streak_axis_type= 'functional')
#Load as UniformDataAxis without warning
s = hs.load('example_hamamatsu.tiff',hamamatsu_streak_axis_type= 'uniform')  
```
After discussing with @jordiferrero we concluded that applying linearisation by default is ok if the user is warned about it.

This will very soon be ready for review, but I'll need to update the docs and change kwds terminology

The merge conflict I have tried to solve numerous time but so far it has escaped my attempts (and understanding)

**Comment by LMSC-NTappy** [comment]
Voilà, now I am ready for a final review!

**Comment by jordiferrero** [comment]
> Ok, now implemented with support for reading non-uniform axes. Behaviour is as follow
> 
> ```python
> s = hs.load('example_hamamatsu.tiff') #Load as a linear axes but warn about linearisation
> #Load as DataAxis
> s = hs.load('example_hamamatsu.tiff',hamamatsu_streak_axis_type= 'data')
> #Load as FunctionalDataAxis (3rd degree polynomial)
> s = hs.load('example_hamamatsu.tiff',hamamatsu_streak_axis_type= 'functional')
> #Load as UniformDataAxis without warning
> s = hs.load('example_hamamatsu.tiff',hamamatsu_streak_axis_type= 'uniform')  
> ```
> 
> After discussing with @jordiferrero we concluded that applying linearisation by default is ok if the user is warned about it.
> 
> This will very soon be ready for review, but I'll need to update the docs and change kwds terminology
> 
> The merge conflict I have tried to solve numerous time but so far it has escaped my attempts (and understanding)

Thanks @LMSC-NTappy ! That's lots of work in this PR.
This PR (with the implementation of `non_linear_axes` made us realise that most loader functions assume linear axis. While this is has always been the standard in hyperspy until now, it means the user is always modifying the raw data upon `hs.load`.
Does this mean that from now onwards, the `hs.load` should always be able to have the non-linear axis argument as done in the example above? Do you think there may be value for letting the user choose if they want the original data axes, a functional one or a linearised one? If you think this should be a point of discussion I can open an Issue :-)

File: doc/user_guide/io.rst [hunk 5]
```diff
@@ -772,17 +772,46 @@ Currently HyperSpy has limited support for reading and saving the TIFF tags.
...
+    linearisation error. Initialise a UniformDataAxis.
+  * if ``data``, the raw time axis parsed from the metadata is used. Initialise
+    a DataAxis.
+  * if ``functional``, the best-fit 3rd-order polynomial axis is used, avoiding
+    linearisation error. Initialise a FunctionalDataAxis.
+
+By default, ``uniform`` is used but a warning of the linearisation error is issued.
+Explicitly passing ``hamamatsu_streak_axis_type='uniform'`` suppresses the warning.
+In all cases, the original axis values are stored in the ``original_metadata`` of the
+signal object.
```
Comment (jordiferrero) [inline doc/user_guide/io.rst:798]: I suppose this is visible in the docstrings of the `hs.load` function? Or is it hidden inside other sub-loading functions?

**Review by jordiferrero — APPROVED** [review]

File: doc/user_guide/io.rst (hunk 5 above)
Comment (LMSC-NTappy) [reply doc/user_guide/io.rst:798]: >  Or is it hidden inside other sub-loading functions?

Exactly this. The load docstring is encyclopedic already, so I think new information would be lost inside. By default, a warning is issued specifying that information directly in the Command Line so  people won't miss it

File: hyperspy/io_plugins/tiff.py [hunk 6]
```diff
@@ -178,41 +208,16 @@ def _read_serie(tiff, serie, filename, force_read_resolution=False,
...
-    axes = [{'size': size,
-             'name': str(name),
-             'scale': scale,
-             'offset': offset,
-             'units': unit,
-             }
-            for size, name, scale, offset, unit in zip(shape, names,
-                                                       scales, offsets,
-                                                       units)]
+        axes = _build_axes_dictionaries(shape, names, scales=[1.0] * len(names),
```
Comment (ericpre) [inline hyperspy/io_plugins/tiff.py]: Maybe it is more simple to set `None` as default value for `names`, `scale`, `offset` and `units` and let `_build_axes_dictionaries` figure out what should be the shape and default values.

File: hyperspy/io_plugins/tiff.py [hunk 7]
```diff
@@ -275,185 +289,483 @@ def _load_data(serie, is_rgb, sl=None, memmap=None, **kwds):
...
+        del op['FEI_SFEG']
+    scales['x'] = float(op['fei_metadata']['Scan']['PixelWidth'])
+    scales['y'] = float(op['fei_metadata']['Scan']['PixelHeight'])
+    units.update({'x': 'm', 'y': 'm'})
+
+    # Finally, axes descriptors can be additionally parsed from digital micrograph or imagej-style files
+    if _is_digital_micrograph(op):
+        scales, offsets, units = _add_axes_digital_micrograph(op, scales, offsets, units)
+    if _is_imagej(tiff):
+        scales, offsets, units = _add_axes_imagej(tiff, op, scales, offsets, units)
```
Comment (ericpre) [inline hyperspy/io_plugins/tiff.py]: Is this necessary? If the file is from FEI, we should use the calibration from the FEI metadata?

File: hyperspy/io_plugins/tiff.py [hunk 8]
```diff
@@ -275,185 +289,483 @@ def _load_data(serie, is_rgb, sl=None, memmap=None, **kwds):
...
+    # can be absent and thus is not used here.
+    scale_in_m = op['CZ_SEM'][''][3] * 1024 / tiff.pages[0].shape[1]
+    scales.update({'x': scale_in_m, 'y': scale_in_m})
+    units.update({'x': 'm', 'y': 'm'})
+
+    # Finally, axes descriptors can be additionally parsed from digital micrograph or imagej-style files
+    if _is_digital_micrograph(op):
+        scales, offsets, units = _add_axes_digital_micrograph(op, scales, offsets, units)
+    if _is_imagej(tiff):
+        scales, offsets, units = _add_axes_imagej(tiff, op, scales, offsets, units)
```
Comment (ericpre) [inline hyperspy/io_plugins/tiff.py]: Same as above

File: hyperspy/io_plugins/tiff.py [hunk 9]
```diff
@@ -275,185 +289,483 @@ def _load_data(serie, is_rgb, sl=None, memmap=None, **kwds):
...
+        _logger.debug("getting TVIPS scale from XYResolution")
+        scales['x'], scales['y'] = _get_scales_from_x_y_resolution(
+            op, factor=1e-2)
         units.update({'x': 'm', 'y': 'm'})
 
+    # Finally, axes descriptors can be additionally parsed from digital micrograph or imagej-style files
+    if _is_digital_micrograph(op):
+        scales, offsets, units = _add_axes_digital_micrograph(op, scales, offsets, units)
+    if _is_imagej(tiff):
+        scales, offsets, units = _add_axes_imagej(tiff, op, scales, offsets, units)
```
Comment (ericpre) [inline hyperspy/io_plugins/tiff.py]: Same as above.

File: hyperspy/io_plugins/tiff.py [hunk 10]
```diff
@@ -275,185 +289,483 @@ def _load_data(serie, is_rgb, sl=None, memmap=None, **kwds):
...
+    op['Olympus_SIS_metadata'] = sis_metadata
+    scales['x'] = round(float(sis_metadata['pixelsizex']), 15)
+    scales['y'] = round(float(sis_metadata['pixelsizey']), 15)
+    units.update({'x': 'm', 'y': 'm'})
+
+    # Finally, axes descriptors can be additionally parsed from digital micrograph or imagej-style files
+    if _is_digital_micrograph(op):
+        scales, offsets, units = _add_axes_digital_micrograph(op, scales, offsets, units)
+    if _is_imagej(tiff):
+        scales, offsets, units = _add_axes_imagej(tiff, op, scales, offsets, units)
```
Comment (ericpre) [inline hyperspy/io_plugins/tiff.py]: Same as above.

File: hyperspy/io_plugins/tiff.py [hunk 11]
```diff
@@ -275,185 +289,483 @@ def _load_data(serie, is_rgb, sl=None, memmap=None, **kwds):
...
+        scales['x'], scales['y'] = _get_scales_from_x_y_resolution(op, factor=scale)
         units = {"x": "1 / nm", "y": "1 / nm", "z": t.Undefined}
-    return scales, units, offsets, intensity_axis
 
-def _get_scales_from_x_y_resolution(op, factor=1):
+    # Finally, axes descriptors can be additionally parsed from digital micrograph or imagej-style files
+    if _is_digital_micrograph(op):
+        scales, offsets, units = _add_axes_digital_micrograph(op, scales, offsets, units)
+    if _is_imagej(tiff):
+        scales, offsets, units = _add_axes_imagej(tiff, op, scales, offsets, units)
```
Comment (ericpre) [inline hyperspy/io_plugins/tiff.py]: Same as above

File: hyperspy/io_plugins/tiff.py [hunk 12]
```diff
@@ -275,185 +289,483 @@ def _load_data(serie, is_rgb, sl=None, memmap=None, **kwds):
...
+        axes = _axes_hamamatsu_streak(tiff, op, shape, names, **kwds)
+        return axes
+    # Axes are otherwise set to defaults
+    else:
+        scales, offsets, units = _axes_defaults()
+        # Axes descriptors can be additionally parsed from digital micrograph or imagej-style files
+        if _is_digital_micrograph(op):
+            scales, offsets, units = _add_axes_digital_micrograph(op, scales, offsets, units)
+        if _is_imagej(tiff):
+            scales, offsets, units = _add_axes_imagej(tiff, op, scales, offsets, units)
```
Comment (ericpre) [inline hyperspy/io_plugins/tiff.py:737]: Move these above at the same level than others (fei, zeiss, etc.)?

**Review by ericpre — COMMENTED** [review]
This looks good. Can you try to compress the two files? You can take example on how to do it from #2926.
See comments about the calibration.

File: hyperspy/io_plugins/tiff.py (hunk 12 above)
Comment (LMSC-NTappy) [reply hyperspy/io_plugins/tiff.py:737]: It seems to me that the tiff loader/writer plugin was written to support digital micrograph or imagej metadata on top of other file formats. 

I think the usecase would be loading an FEI file with FEI axes calibrations, but with some calibration aspects, e.g. one axis or a noise model (see _test_read_unit_from_dm in the test_tiff.py for eyample), calibrated using dm3 or imagej-style metadata.

The problem is that not all the metadata parsers are supporting this, so I found it difficult to address this in a single level of code...

File: hyperspy/io_plugins/tiff.py (hunk 6 above)
Comment (LMSC-NTappy) [reply hyperspy/io_plugins/tiff.py]: Will try to do it

File: hyperspy/io_plugins/tiff.py (hunk 12 above)
Comment (ericpre) [reply hyperspy/io_plugins/tiff.py:737]: > It seems to me that the tiff loader/writer plugin was written to support digital micrograph or imagej metadata on top of other file formats.
>
> I think the usecase would be loading an FEI file with FEI axes calibrations, but with some calibration aspects, e.g. one axis or a noise model (see _test_read_unit_from_dm in the test_tiff.py for eyample), calibrated using dm3 or imagej-style metadata.

I am not sure, it may simply have been changed over time in an inconsistent manner...
 
> The problem is that not all the metadata parsers are supporting this, so I found it difficult to address this in a single level of code...

This is why, I would remove it.

**Comment by ericpre** [comment]
@LMSC-NTappy, when do you think you will have time to finish it?

**Comment by LMSC-NTappy** [comment]
> @LMSC-NTappy, when do you think you will have time to finish it?

Earliest I can do is to free some time on tuesday to wrap it up. Otherwise when I am back from a conference on may the 2nd

**Comment by jlaehne** [comment]
Thanks Eric for finishing this one up!
